### PR TITLE
WIP: Add Boost

### PR DESCRIPTION
**Pull request by jakirkham**
Builds Boost from source. This is largely borrowed from conda-recipes.

However, we build Boost on Mac and Linux with C++11 support. Not doing this has resulted in issues with other C++11 compiled libraries on Mac that require Boost support. However, we have not found this to be a problem when intermixing libraries that don't use C++11. Also, this means that we use `clang` and `libc++` on Mac. That all being said, we are able to keep compatibility with Mac 10.7. Maybe this will be controversial so I am laying it out here so we can discuss. IMHO this is the right way forward as `gcc` on Mac is ancient and support for it ends up being removed.

This strategy cannot easily be applied to Windows as C++11 only exists in limited form on older Windows compilers. Though if someone with more expertise in that area would like to share a possible approach forward, I am open to discussion.

### CODE REVIEW

File: recipes/boost/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+source:
+  fn:  boost_{{ version.replace(".", "_") }}.tar.bz2
+  url: http://sourceforge.net/projects/boost/files/boost/{{ version }}/boost_{{ version.replace(".", "_") }}.tar.bz2
+  md5: 65a840e1a0b13a558ff19eeb2c4f0cbe
+
+build:
+  features:
+    - vc9               [win and py27]
+    - vc10              [win and py34]
+    - vc14              [win and py35]
```
Comment (ocefpaf) [inline recipes/boost/meta.yaml]: I guess that the new linter will enforce the comment style (`#`) here.

File: recipes/boost/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{% set version = "1.60.0" %}
+
+package:
+  name: boost
+  version: {{ version }}
+
+source:
+  fn:  boost_{{ version.replace(".", "_") }}.tar.bz2
+  url: http://sourceforge.net/projects/boost/files/boost/{{ version }}/boost_{{ version.replace(".", "_") }}.tar.bz2
```
Comment (ocefpaf) [inline recipes/boost/meta.yaml]: Moments like this I start to wonder if the jinja var is a good idea. I am not against it, but it sounds more trouble than it is worth. (I never really like those vars as the URL becomes impossible to copy-and-paste :wink:)

**Comment by jakirkham** [comment]
If any of the original authors (@ccordoba12 @patricksnape @frol @mutirri) have feedback on this, that would be great. Also, if any of you would be willing to be added as maintainers that would be awesome, please let me know and I will add you.

File: recipes/boost/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: We could assign `version.replace(".", "_")` to another variable if it would be a bit nicer.

File: recipes/boost/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Yes, sorry, slipped pass me. Will fix that. Thanks.

Comment (jakirkham) [inline recipes/boost/meta.yaml]: Fixed. :smile:

File: recipes/boost/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: So, I refactored out a `filename` variable.

File: recipes/boost/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{% set version = "1.60.0" %}
+{% set filename = "boost_%s.tar.bz2" % version.replace(".", "_") %}
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Here is the filename variable.

File: recipes/boost/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+package:
+  name: boost
+  version: {{ version }}
+
+source:
+  fn:  {{ filename }}
+  url: http://sourceforge.net/projects/boost/files/boost/{{ version }}/{{ filename }}
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Now, `fn` and `url` are still readable while being maintainable. :smile: Thoughts?

**Comment by frol** [comment]
It looks good to me.

File: recipes/boost/meta.yaml (hunk 4 above)
Comment (patricksnape) [inline recipes/boost/meta.yaml]: Looks good to me!

**Comment by patricksnape** [comment]
The issue that I was having with Boost was trying to get all versions of Python to build with it - specifically on Windows. I had to make a custom JAM file to get it to work - though I'm not sure if this is still required with the latest boost.

File: recipes/boost/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+test:
+  commands:
+    - test -f $PREFIX/lib/libboost_atomic.dylib       # [osx]
+    - test -f $PREFIX/lib/libboost_atomic.so          # [linux]
+    - test -f $PREFIX/lib/libboost_chrono.a           # [unix]
+    - test -f $PREFIX/lib/libboost_chrono.dylib       # [osx]
+    - test -f $PREFIX/lib/libboost_chrono.so          # [linux]
+    - test -f $PREFIX/lib/libboost_container.dylib    # [osx]
+    - test -f $PREFIX/lib/libboost_container.so       # [linux]
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: This is only a small selection of the Boost libraries. There are many more to add.

**Comment by jakirkham** [comment]
I believe that @ccordoba12 did quite a bit of work in order to get the Windows portion to work. So, that should actually be working now, but CI will verify.

Also, @patricksnape, I know I owe you a longer reply on the discussion thread. It is coming (partially written). :smile:

File: recipes/boost/meta.yaml (hunk 5 above)
Comment (patricksnape) [inline recipes/boost/meta.yaml]: Maybe use one of those fancy jinja variables for the extension?

File: recipes/boost/bld.bat [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    variant=release ^
+    threading=multi ^
+    link=shared ^
+    -j%CPU_COUNT% ^
+    -s ZLIB_INCLUDE="%LIBRARY_INC%" ^
+    -s ZLIB_LIBPATH="%LIBRARY_LIB%"
+if errorlevel 1 exit 1
+
+:: Install fix-up for a non version-specific boost include
+move %LIBRARY_INC%\boost-1_60\boost %LIBRARY_INC%
```
Comment (patricksnape) [inline recipes/boost/bld.bat]: Hardcoded boost version? Would it be possible to pass the boost version through ala:

```
build:
  script_env:
   - BOOST_VER={{ version }}
```

Comment (jakirkham) [inline recipes/boost/bld.bat]: Good point or maybe we could use some regex type thing on Windows. Not positive on the syntax though. So, if you have some ideas, we can play with that a bit.

File: recipes/boost/bld.bat [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+if errorlevel 1 exit 1
+
+:: Move dll's to LIBRARY_BIN
+move %LIBRARY_LIB%\*vc%LIB_VER%-mt-1_60.dll "%LIBRARY_BIN%"
```
Comment (jakirkham) [inline recipes/boost/bld.bat]: Will want it here too.

File: recipes/boost/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Cool, glad we like this. If you guys have more thoughts on how to make things more readable still, please share.

File: recipes/boost/meta.yaml (hunk 5 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Yes, I was think that exact same point. Otherwise this is going to be a massive pain. That or just have test scripts and offload that logic to bash and batch.

Do you know what the Windows situation looks like here? I have a sort of working Windows computer now so I can play with this locally too.

Comment (jakirkham) [inline recipes/boost/meta.yaml]: Also, think I may need `lib64` in some cases here too.

**Comment by jakirkham** [comment]
Thanks for the feedback on this thus far guys. This is great. :smile:

Will keep working on it to incorporate your suggestions. Keep them coming though. Definitely helpful.

File: recipes/boost/bld.bat (hunk 6 above)
Comment (ccordoba12) [inline recipes/boost/bld.bat]: I know it's ugly to have this hardcoded version here (I'm the culprit for this one ;-). As @patricksnape says, it should be easily fixable with some Jinja2 magic.

Comment (jakirkham) [inline recipes/boost/bld.bat]: No worries. Getting the first working implementation is the hard part. Thanks for getting that together.

Now we will just make it pretty. :wink:

File: recipes/boost/meta.yaml (hunk 5 above)
Comment (ccordoba12) [inline recipes/boost/meta.yaml]: This test needs to be as comprehensible as possible because Boost build system *never* fails, it just skips the libraries it can't compile.

Please include Windows dll's here too :-)

**Comment by ccordoba12** [comment]
Great work guys!! This will simplify my life so much! :-)

Please list me as a maintainer.

**Comment by jakirkham** [comment]
Thanks for your work on getting the cross platform support work. This is really nice. Thanks.

> Please list me as a maintainer.

Excellent. Will do. Thanks.

**Comment by ccordoba12** [comment]
> However, we build Boost on Mac and Linux with C++11 support

Wise decision. We haven't been able to support C++11 at Continuum because we haven't settled in a new gcc toolchain.

> This strategy cannot easily be applied to Windows as C++11 only exists in limited form on older Windows compilers

That's right. Full C++11 support is only available in Visual Studio 2012. 2010 has some limited support for it.

**Comment by jakirkham** [comment]
@msarahan added some nice additions for MSVC ( https://github.com/jakirkham/staged-recipes/pull/1 ), which have been added here. Any Windows people please feel free to give feedback on this.

Also, @ukoethe here is the main Boost PR. Are there any other options that you are using to get VIGRA to work well with Boost? I would like this to be easily usable with it and plan to get VIGRA 1.11.0 (thanks for the release) in next.

**Comment by msarahan** [comment]
From @ukoethe in https://github.com/jakirkham/staged-recipes/pull/1

> Yes, I do a similar thing in my recipes. However, to avoid replication of such code in every recipe, I have the visual-studio recipe install a little script that provides utility functions and variables for other recipes to call, see https://github.com/ilastik/ilastik-build-conda/blob/toolset-config/visual-studio/common-vars.bat.in

This is a much, much nicer way to do it.  I shy away from it because I feel like it truly belongs in a place that will always be there (like conda-build itself).  If we have recipes that use this approach, but the tools are part of conda-forge's internals, then what if someone wants to use recipes hosted on conda-forge?

I'll see about putting it into conda-build.  (not for this PR though)

**Comment by ccordoba12** [comment]
@jakirkham, the build failure on Linux is caused because Boost 1.60 has compilation issues with gcc 4.4. We need to use a more recent gcc for that.

**Comment by jakirkham** [comment]
> That's right. Full C++11 support is only available in Visual Studio 2012. 2010 has some limited support for it.

Could we maybe add C++0x experimental support for Windows? Do you have any thoughts on this? I don't know how much it helps, but maybe it would be a little more similar.

**Comment by msarahan** [comment]
When I played with that stuff a while back, it was extremely painful
because all the C++0x stuff was in a separate namespace.  It required quite
a bit of hacking on every project to use the extra namespace.  I'm not sure
that's an easy solution for us as package builders.

**Comment by jakirkham** [comment]
> ...the build failure on Linux is caused because Boost 1.60 has compilation issues with gcc 4.4. We need to use a more recent gcc for that.

Right. Thanks, @ccordoba12. I suppose we could temporarily disable C++11 features on Linux. Once the new docker container with the latest gcc gets integrated then we can re-enable them. I'd really like C++11, but I don't want to hold this up over that or block the use of Boost on Linux.

**Comment by jakirkham** [comment]
Do you know if @mingwandroid has any thoughts on getting C++11 support on Windows?

File: recipes/boost/meta.yaml (hunk 5 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: > ...Boost build system never fails, it just skips the libraries it can't compile.

:laughing: Don't I know it. Have been bitten by this before.

> Please include Windows dll's here too :-)

Absolutely, I just wasn't clear on the naming scheme when I added this. Will check on my Windows machine and update.

**Comment by ukoethe** [comment]
> I'll see about putting it into conda-build.

I'm not sure that conda-build is the right place, because these scripts differ for each Visual Studio version. This has an important advantage: when I install a particular version of the `visual-studio` meta-package, I can check that the VC version in the PATH is the same as the one in the meta-package. This has saved me from a lot of stupid mistakes, for example when I attempted to start a vc14 build in a vc11 console, or a 64-bit build with a 32-bit compiler. Thanks to the script, these builds fail right away.

**Comment by ukoethe** [comment]
> Full C++11 support is only available in Visual Studio 2012.

That's unfortunately not true. Full support is available in Visual Studio 2015 and probably more or less in Visual Studio 2013 (I haven't tried the latter, though). Visual Studio 2012 supports a decent subset, but lacks variadic templates and initializer lists, for example.

File: recipes/boost/bld.bat [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+:FOUND
+
+call :TRIM VSTRING %VSTRING%
+
+:: Start with bootstrap
+call bootstrap.bat
+if errorlevel 1 exit 1
+
+:: Build step
+.\b2 install ^
```
Comment (ukoethe) [inline recipes/boost/bld.bat]: I'm not sure if `--with-python` is needed here to have the boost_python library built. Please check!

File: recipes/boost/bld.bat (hunk 6 above)
Comment (ukoethe) [inline recipes/boost/bld.bat]: It's easy to fix this in the bat script. But since the version number was only recently added to the boost include path (it wasn't there in 1.55), chances are that they change their mind again in 1.61. So I suggest to wait until we see the next version.

File: recipes/boost/bld.bat (hunk 8 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: Good point. Probably should be.

Comment (jakirkham) [inline recipes/boost/bld.bat]: Added.

**Comment by jakirkham** [comment]
Would it be alright if we use the Anaconda `gcc` compiler for Linux here? That way we can have C++11 support on Linux. Was that compiler built on CentOS 5.11?

**Comment by msarahan** [comment]
For now, I'd say sure.  Once we sort out the docker build image question, this will not be necessary.  We can remove it then.

I didn't build that compiler, but I think it was built on CentOS 5.

File: recipes/boost/bld.bat (hunk 8 above)
Comment (ukoethe) [inline recipes/boost/bld.bat]: As I think about it, I can't remember which libraries boost includes by default on Windows - maybe none? So, perhaps `--with-serialization --with-system` and so on is needed for each of them.

**Comment by jakirkham** [comment]
Cool, let's give that a go then. Absolutely, we will strip it out once we have the docker gcc figured out.

File: recipes/boost/bld.bat (hunk 8 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: You may be right. Let me work on the test part. I'll make sure this does a comprehensive test on the libs for all platforms. That way, it will be clear what (if any) other flags need to be added here.

File: recipes/boost/bld.bat [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+echo @echo 18=12 >> msvc_versions.bat
+echo @echo 19=14 >> msvc_versions.bat
+
+REM Run cl.exe to find which version our compiler is
+for /f "delims=" %%A in ('cl /? 2^>^&1 ^| findstr /C:"Version"') do set "CL_TEXT=%%A"
+FOR /F "tokens=1,2 delims==" %%i IN ('msvc_versions.bat') DO echo %CL_TEXT% | findstr /C:"Version %%i" > nul && set VSTRING=%%j && goto FOUND
+EXIT 1
+:FOUND
+
+call :TRIM VSTRING %VSTRING%
```
Comment (patricksnape) [inline recipes/boost/bld.bat]: This is truly black magic to me. But it looks fantastic. How does this ensure that the 'correct' compiler is chosen for normal builds of Python? Aka VC9 for PY27, VC10 for PY34, VC14 for PY35?

Comment (msarahan) [inline recipes/boost/bld.bat]: conda-build takes care of that.  The change I asked @jakirkham to make basically gets out of the way if the user wants to break that relationship.  Conda-build calls the vcvarsall.bat of the compiler you tell it (or the defaults line up with Python versions).  It does that before bld.bat is called.  This script uses whichever cl.exe is on PATH to figure out the corresponding VS version.

I have no idea why that for loop mess is necessary to read output from programs, but that is all I've been able to find.  I feel dirty every time I do that (and I also have to spend a fair bit of time on Stack Overflow every time I add a new variant).

Comment (jakirkham) [inline recipes/boost/bld.bat]: Yeah, I squashed all this stuff together to keep life simple. The original change can be found in this PR ( https://github.com/jakirkham/staged-recipes/pull/1 ).

File: recipes/boost/bld.bat [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    -s ZLIB_INCLUDE="%LIBRARY_INC%" ^
+    -s ZLIB_LIBPATH="%LIBRARY_LIB%"
+if errorlevel 1 exit 1
+
+:: Install fix-up for a non version-specific boost include
+move %LIBRARY_INC%\boost-1_60\boost %LIBRARY_INC%
+if errorlevel 1 exit 1
+
+:: Move dll's to LIBRARY_BIN
+move %LIBRARY_LIB%\*vc%VSTRING%0-mt-1_60.dll "%LIBRARY_BIN%"
```
Comment (jakirkham) [inline recipes/boost/bld.bat]: If we do templating, we will want it here too.

**Comment by jakirkham** [comment]
Would you like to be added as a maintainer, @msarahan?

**Comment by msarahan** [comment]
Sure, I'll be happy to help.

**Comment by jakirkham** [comment]
Thanks!

File: recipes/boost/bld.bat (hunk 8 above)
Comment (ccordoba12) [inline recipes/boost/bld.bat]: `--with-python` is not necessary, nor the other options mentioned by @ukoethe. As I said before, Boost build system tries to compile everything it can, and it skips the libraries it can't.

@jakirkham, please remove `--with-python` from here.

**Comment by ukoethe** [comment]
@jakirkham Please add a patch to silence the stupid deprecation warnings (boost maintainers marked functions deprecated that are still called within boost itself). See https://github.com/ilastik/ilastik-build-conda/blob/toolset-config/boost/1.60.0.patch

**Comment by ukoethe** [comment]
Reviewing the patch linked above, I noticed that auto-linking doesn't work properly for Python 3 on Windows, because the name of the boost_python library is inconsistent between header and build system, leading to "file not found" errors during linking. The patch fixes this as well.

File: recipes/boost/bld.bat (hunk 8 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: Noted will remove.

Comment (jakirkham) [inline recipes/boost/bld.bat]: I think I mentioned this somewhere else, but ensuring we have verified that we have *all* the libraries in the test will make it easy to see if something is missing and then we can add flags as needed.

**Comment by jakirkham** [comment]
Could you please explain where these warning are coming from, @ukoethe? Is this just during the build or during runtime? If it is the former, I am inclined to leave it alone.

File: recipes/boost/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    - test -f $PREFIX/lib/libboost_regex.so                   # [linux]
+    - test -f $PREFIX/lib/libboost_serialization.so           # [linux]
+    - test -f $PREFIX/lib/libboost_signals.so                 # [linux]
+    - test -f $PREFIX/lib/libboost_system.so                  # [linux]
+    - test -f $PREFIX/lib/libboost_thread.so                  # [linux]
+    - test -f $PREFIX/lib/libboost_timer.so                   # [linux]
+    - test -f $PREFIX/lib/libboost_type_erasure.so            # [linux]
+    - test -f $PREFIX/lib/libboost_unit_test_framework.so     # [linux]
+    - test -f $PREFIX/lib/libboost_wave.so                    # [linux]
+    - test -f $PREFIX/lib/libboost_wserialization.so          # [linux]
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: This is pretty gross as it is soooo long. Sorry, many libraries. However, this makes it trivial to add more when they show up and it is clear to anyone what is going on here.

Only have done Mac and Linux. If someone can point me to the equivalent syntax for Windows, I can add that now to.

File: recipes/boost/bld.bat (hunk 8 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: `--with-python` has been removed.

File: recipes/boost/meta.yaml (hunk 11 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Thinking about tidying this with jinja just so it stays readable.

**Comment by ukoethe** [comment]
> Could you please explain where these warning are coming from, @ukoethe? 

These warning are triggered when you use boost_python as a dependency. For example, the VIGRA build is full of them, although VIGRA never calls these functions directly.

File: recipes/boost/meta.yaml (hunk 11 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Would something like this work for Windows?

    if not exist <filename> exit 1

**Comment by jakirkham** [comment]
> For example, the VIGRA build is full of them, although VIGRA never calls these functions directly.

This isn't actually showing up during `import vigra` though, correct?

**Comment by ukoethe** [comment]
> This isn't actually showing up during import vigra though, correct?

No, but it's cluttering vigra's build output (pages of useless warnings), and genuine problems are shadowed. Without the patch, boost 1.60.0 is pretty unusable for development.

File: recipes/boost/meta.yaml (hunk 11 above)
Comment (ukoethe) [inline recipes/boost/meta.yaml]: Yes, this works.

Comment (jakirkham) [inline recipes/boost/meta.yaml]: Thanks. I'll update the Windows tests and give this a try.

File: recipes/boost/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            "timer",
+            "type_erasure",
+            "unit_test_framework",
+            "wave",
+            "wserialization"
+    ] %}
+    {% for each_boost_lib in boost_libs %}
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.dylib  # [osx]
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.so     # [linux]
+    {% endfor %}
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Now has a nice clean jinja template based implementation. Still quite long as we have to explicitly list all the library names. However, with this strategy we don't need to explicitly list every library for every system. Instead we have library names for all systems and then system naming conventions. Tried it out on my Mac and it worked. Will now add Windows.

File: recipes/boost/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            "unit_test_framework",
+            "wave",
+            "wserialization"
+    ] %}
+    {% for each_boost_lib in boost_libs %}
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.dylib                                         # [osx]
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.so                                            # [linux]
+    - if not exist $PREFIX/Library/lib/boost_{{ each_boost_lib }}-vc90-mt-{{ win_vstr }}.lib exit 1   # [win and py27]
+    - if not exist $PREFIX/Library/lib/boost_{{ each_boost_lib }}-vc100-mt-{{ win_vstr }}.lib exit 1  # [win and py34]
+    - if not exist $PREFIX/Library/lib/boost_{{ each_boost_lib }}-vc140-mt-{{ win_vstr }}.lib exit 1  # [win and py35]
```
Comment (ccordoba12) [inline recipes/boost/meta.yaml]: We also should test for dll's presence, shouldn't we?

File: recipes/boost/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  features:
+    - vc9               # [win and py27]
+    - vc10              # [win and py34]
+    - vc14              # [win and py35]
+
+requirements:
+  build:
+    - gcc               # [linux]
+    - python
+    - icu               # [unix]
```
Comment (ccordoba12) [inline recipes/boost/meta.yaml]: `icu` version must be pinned, to avoid errors like ContinuumIO/anaconda-issues#700.

Comment (ccordoba12) [inline recipes/boost/meta.yaml]: That was my fault: I compiled a more recent version of icu for my Qt5 packages, and forget about it when creating new Boost packages.

That's why it has to be pinned :-)

Comment (msarahan) [inline recipes/boost/meta.yaml]: But this opens Pandora's box, right?  What pinning should be done in general?

Comment (jakirkham) [inline recipes/boost/meta.yaml]: Could use selectors so it is only pinned on Windows.

File: recipes/boost/meta.yaml (hunk 13 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: So, use to Mac and Linux I didn't even think they would be libraries not in `lib`. :smile: Should be fixed, along with several other errors, in the most recent commit.

File: recipes/boost/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    ] %}
+    {% for each_boost_lib in boost_libs %}
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.dylib                                             # [osx]
+    - test -f $PREFIX/lib/libboost_{{ each_boost_lib }}.so                                                # [linux]
+    - if not exist %PREFIX%\\Library\\bin\\boost_{{ each_boost_lib }}-vc90-mt-{{ win_vstr }}.dll exit 1   # [win and py27]
+    - if not exist %PREFIX%\\Library\\bin\\boost_{{ each_boost_lib }}-vc100-mt-{{ win_vstr }}.dll exit 1  # [win and py34]
+    - if not exist %PREFIX%\\Library\\bin\\boost_{{ each_boost_lib }}-vc140-mt-{{ win_vstr }}.dll exit 1  # [win and py35]
+    - if not exist %PREFIX%\\Library\\lib\\boost_{{ each_boost_lib }}-vc90-mt-{{ win_vstr }}.lib exit 1   # [win and py27]
+    - if not exist %PREFIX%\\Library\\lib\\boost_{{ each_boost_lib }}-vc100-mt-{{ win_vstr }}.lib exit 1  # [win and py34]
+    - if not exist %PREFIX%\\Library\\lib\\boost_{{ each_boost_lib }}-vc140-mt-{{ win_vstr }}.lib exit 1  # [win and py35]
```
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Can someone with more Windows knowledge clarify as to whether this looks correct?

Comment (msarahan) [inline recipes/boost/meta.yaml]: LGTM.  Let's see what Appveyor says.

File: recipes/boost/meta.yaml (hunk 14 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: At some point, we do need to consider pinning to ensure ABI compatibility with regards to non-Python and non-NumPy things. Though this requires a more general proposal that I believe @ukoethe was thinking about.

File: recipes/boost/bld.bat (hunk 10 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: Does anyone know if I can share jinja template variables with the environment? That would help fix these hard-coded versions and leverage the same info in the recipe. Alternatively, if we can just use `*`, I can do that here also.

File: recipes/boost/meta.yaml (hunk 15 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: The first AppVeyor build in the matrix passed and did actually check the libraries. So, that is a good sign. Will be a bit before the full outcome is clear.

**Comment by jakirkham** [comment]
@ukoethe, I understand your concern with the Boost build warnings. Mainly I am just trying to ensure that I have a clear idea of the scope of the problem.

Our general policy with patches has been that we submit them upstream. We want to avoid being in a situation where we are maintaining our own special fork of packages here. Not to say this small patch makes this the case, but that not proposing patches upstream does move us further in this direction. So, this is a step we will need to take with this patch, as well. This doesn't mean that the patch needs to be accepted as is by Boost. Just that it must be proposed via the mechanism they would prefer and followed up on periodically.

Though I expect they will want to fix this problem. They may not take the patch as is, but instead may want to make modifications or solve this in another manner. As this could become a longer discussion (give the reach of the Boost project), we are going to want to be able to track the progress on this issue with Boost. Personally, I think this will be much easier to do on the Boost feedstock itself (once generated) instead of having it also co-mingled with this 70+ comment PR. In other words, we need to make it easy to track progress on that problem as it will be hard enough to follow anyways.

In short, I agree with you that it is a problem and agree with you that we want it fixed. Given the nature of the problem, we can convert this to a feedstock and make the first build without it, but incorporate it after that has happened. I'm not saying we won't take the patch, but that I am not ready to take it yet just so that we can stay a bit better organized. I hope this makes sense.

**Comment by jakirkham** [comment]
Alright, I'd like to take the general temperature on this PR.

I understand that as this is a very useful and widely applicable package. There is a lot of interest in it and everyone has different needs and expectations from it. That being said, I don't think we are going to solve all of the problems here. This will remain something that sees continually development and maintenance at conda-forge. So, we have to make a decision at some point as to what constitutes done enough to create the first build.

How do people feel about merging this (given CI passes)? What sort of concerns remain? Instead of commenting on a bunch of lines, maybe a brief comment here will help me get an idea of what your thoughts are. Thanks in advance.

Also, thanks everyone for your constructive feedback.

File: recipes/boost/meta.yaml (hunk 14 above)
Comment (jakirkham) [inline recipes/boost/meta.yaml]: Now that I am thinking about it. I think the easiest way forward on this would be to get an `icu` recipe going here.

Comment (msarahan) [inline recipes/boost/meta.yaml]: +1.  We have a recipe I can convert - don't redo it if you don't have to.

**Comment by msarahan** [comment]
+1 for merge from me.

I think a good "first build" is that it:
  - passes the linters
  - passes all reasonably achievable platforms
  - implements some meaningful test of functionality
  - documents any required patches

File: recipes/boost/meta.yaml (hunk 14 above)
Comment (msarahan) [inline recipes/boost/meta.yaml]: https://github.com/conda-forge/staged-recipes/pull/198

Comment (ccordoba12) [inline recipes/boost/meta.yaml]: `icu` is critical to Boost, as the issue I referenced above shows. It seems `icu` changes its library version name on each release, so that's why it needs to be pinned with something like:

    - icu 54*

I don't understand Mike concern because (for example) we ping `libpng` in our Qt recipe with

    - libpng      1.6*

to avoid breaking Qt when libpng is updated :-)

**Comment by ccordoba12** [comment]
+1 from me too, but after pinning `icu`. I think that's really important :-)

File: recipes/boost/meta.yaml (hunk 14 above)
Comment (msarahan) [inline recipes/boost/meta.yaml]: I don't mind pinning, I would just like to have completely clear guidelines on when and why to do it.  I have seen things fail because versions were not pinned, and I've seen things fail because things were pinned.  Neither approach is completely right, but people should be making explicit, educated choices on these things.

File: recipes/boost/meta.yaml (hunk 4 above)
Comment (pelson) [inline recipes/boost/meta.yaml]: The filename could be defined as a jinja variable *after* the ``source:`` section? That way it is in one place, readable, and doesn't pollute the sacred first few lines... otherwise, I :heart: it

**Comment by pelson** [comment]
Super review everybody. :+1: on this.

Regarding pinning - I agree that it needs doing, but also that it is opening pandora's box. I've attempted a few times to make pinning easier from within a recipe (within conda-build) to no avail (https://github.com/conda/conda-build/pull/747, https://github.com/conda/conda-build/pull/650) - that said, there is nothing stopping us hardcoding the icu version into the recipe. I think this is a discussion we should take with us into https://github.com/conda-forge/boost-feedstock.

**Comment by ukoethe** [comment]
> @jakirkham Our general policy with patches has been that we submit them upstream.

Of course, we should submit it upstream. Please find someone to take care of it. However, since 1.60 is pretty new, it may take a while for the changes to take effect (experiences with PRs on boost_python have not been too encouraging). I'd like a usable boost package now. 

Also don't forget the autolinking patch I pointed out earlier - under certain conditions, this bug is a real showstopper .

**Comment by ccordoba12** [comment]
@jakirkham, we really shouldn't be so picky about patches. What we really want is something that's useful for our users (at least that's our policy at Continuum :-)

@ukoethe, please submit your patches to the Boost feedstock and I'll merge them in my maintainer capacity. I also have a patch of my own :-)

**Comment by jakirkham** [comment]
> we really shouldn't be so picky about patches. What we really want is something that's useful for our users (at least that's our policy at Continuum :-)

Again never said not to the patch. Just to be clear.

What I did say was no to adding it here. Just think with a project like Boost there will be a fair, but of cross referencing and dialog to get this fixed as it should be (in Boost). I wanted to make sure we didn't have to navigate this long PR as part of the process. Just trying to keep the maintenance burden low. :wink:

**Comment by ccordoba12** [comment]
> Just trying to keep the maintenance burden low.

You're right, and sorry if I misunderstood you :-) We can merge each patch in a feedstock PR, and that's going to be much simpler and cleaner!

**Comment by jakirkham** [comment]
No worries. It was my bad having such a long response to this point. We are all very busy people. :smile:

Here is the [feedstock]( https://github.com/conda-forge/boost-feedstock ), @ukoethe.

**Comment by jakirkham** [comment]
@jschueller @yungyuc, we are adding `boost` to [conda-forge]( https://conda-forge.github.io/ ). This is the community packaging system for use with [`conda`]( http://conda.pydata.org/docs/ ). Packages built here are built with CI and deployed to the [conda-forge channel]( https://anaconda.org/conda-forge/ ) automatically after committing on `master` of a feedstock (a repo with the recipe plus automatically maintained CI scripts). Eventually, packages submitted here will be included in the `defaults` channel after some auditing process as [mentioned]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ). If there is anything here that you see is off or needs to be fixed, please let me know. If you would like to have access to this [feedstock/recipe]( http://github.com/conda-forge/hdf5-feedstock/ ), I would happily add you to the maintainers list so you can have commit privileges for this recipe. This would be useful for updating the version for instance.

File: recipes/boost/bld.bat (hunk 6 above)
Comment (jakirkham) [inline recipes/boost/bld.bat]: If someone wants to submit a PR to the feedstock for this change, feel free to.

Comment (jakirkham) [inline recipes/boost/bld.bat]: BTW, I tried to add functionality like this to `conda-build` once ( see this PR https://github.com/conda/conda-build/pull/767 ), but there were no takers. So, I don't believe this is possible at present.

Comment (jakirkham) [inline recipes/boost/bld.bat]: Though it may be more practical for this problem to just use [`PKG_VERSION`]( http://conda.pydata.org/docs/building/environment-vars.html#environment-variables-set-during-the-build-process ).
